File: Crio/crio_programming_interview_problems-master/SpiralMatrix/SpiralMatrix.java
import java.io.*;
import java.util.*;

class SpiralMatrix  {
  // TODO: CRIO_TASK_MODULE_L2_PROBLEMS
  // complete the below function implementation

  public List<Integer> spiralOrder(int[][] matrix) {
    List<Integer> lst = new ArrayList<Integer>();
    // CRIO_SOLUTION_START_MODULE_L2_PROBLEMS
    int n = matrix.length;
    if (n == 0) {
      return lst;
    }
    int m = matrix[0].length;

    int rStart = 0;
    int cStart = 0;
    int rEnd = n - 1;
    int cEnd = m - 1;

    do {

      //hor - >forward
      for (int i = cStart; i <= cEnd; ++i) {
        lst.add(matrix[rStart][i]);
      }
      // ver -> down
      for (int j = rStart + 1; (j <= rEnd - 1); ++j) {
        lst.add(matrix[j][cEnd]);
      }

      // hor - back
      for (int i = cEnd; (i >= cStart); --i) {
        lst.add(matrix[rEnd][i]);
      }

      // ver - up
      for (int j = rEnd-1; (j >= rStart + 1); --j) {
        lst.add(matrix[j][cStart]);
      }

      rStart += 1;
      cStart += 1;
      rEnd -= 1;
      cEnd -= 1;

    }while (rStart <= rEnd && cStart <= cEnd);

    return lst.subList(0, n * m);
    // CRIO_SOLUTION_END_MODULE_L2_PROBLEMS
    // CRIO_UNCOMMENT_START_MODULE_L2_PROBLEMS
    // return lst;
    // CRIO_UNCOMMENT_END_MODULE_L2_PROBLEMS
  }

  public static void main(String args[]) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();
    int m = scanner.nextInt();
    int[][] matrix = new int[n][m];

    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < m; ++j) {
        matrix[i][j] = scanner.nextInt();
      }
    }
    scanner.close();
    List<Integer> result = new SpiralMatrix().spiralOrder(matrix);

    for (int i = 0; i < result.size(); ++i) {
      System.out.printf("%d ", result.get(i));
    }
  }
}

File: Crio/crio_programming_interview_problems-master/SpiralMatrix/SpiralMatrix.cpp
#include "FastIO.hpp"
#include "PrintMatrix.hpp"
#include "ReadMatrix.hpp"
#include <bits/stdc++.h>
using namespace std;

class SpiralMatrix {
  public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        // TODO: CRIO_TASK_MODULE_L2_PROBLEMS
        // Your implementation goes here
        // CRIO_SOLUTION_START_MODULE_L2_PROBLEMS
        int szy = matrix.size();
        if (szy == 0)
            return res;
        int szx = matrix[0].size();
        if (szx == 0)
            return res;
        vector<bool> bool_vec(szx, false);
        vector<vector<bool>> visit(szy, bool_vec);
        int vx = 1;
        int vy = 0;
        int tmp;
        int cur_x = 0;
        int cur_y = 0;
        while (true) {
            res.push_back(matrix[cur_y][cur_x]);
            visit[cur_y][cur_x] = true;
            if (cur_x + vx == -1 || cur_x + vx == szx || cur_y + vy == -1 ||
                cur_y + vy == szy || visit[cur_y + vy][cur_x + vx]) {
                tmp = vx;
                vx = -vy;
                vy = tmp;
            }
            cur_x += vx;
            cur_y += vy;
            if (cur_y == -1 || cur_y == szy || cur_x == -1 || cur_x == szx ||
                visit[cur_y][cur_x])
                break;
        }
        // CRIO_SOLUTION_END_MODULE_L2_PROBLEMS
        return res;
    }
};

int main() {
    FastIO();
    int n, m;
    cin >> n >> m;
    vector<vector<int>> matrix;
    ReadMatrix<int>().TwoDMatrix(n, m, matrix);
    vector<int> result = SpiralMatrix().spiralOrder(matrix);
    PrintMatrix<int>().OneDMatrix(result, " ");
    return 0;
}
